src/libsystemd/sd-varlink/sd-varlink.c

```diff
@@ -4193,27 +4204,55 @@ _public_ int sd_varlink_error_to_errno(const char *error, sd_json_variant *param
                 { SD_VARLINK_ERROR_EXPECTED_MORE,          -EBADE         },
         };
 
+        int r;
+
         if (!error)
                 return 0;
 
         FOREACH_ELEMENT(t, table)
                 if (streq(error, t->error))
                         return t->value;
 
-        if (streq(error, SD_VARLINK_ERROR_SYSTEM) && parameters) {
-                sd_json_variant *e;
+        /* This following tries to reverse the operation sd_varlink_error_errno() applies to turn errnos into
+         * varlink errors */
+        if (!streq(error, SD_VARLINK_ERROR_SYSTEM))
+                return -EBADR;
 
-                e = sd_json_variant_by_key(parameters, "errno");
-                if (sd_json_variant_is_integer(e)) {
-                        int64_t i;
+        if (!parameters)
+                return -EBADR;
 
-                        i = sd_json_variant_integer(e);
-                        if (i > 0 && i < ERRNO_MAX)
-                                return -i;
-                }
+        /* If an origin is set, check if it's Linux, otherwise don't translate */
+        sd_json_variant *e = sd_json_variant_by_key(parameters, "origin");
+        if (e && (!sd_json_variant_is_string(e) ||
+                  !streq(sd_json_variant_string(e), "linux")))
+                return -EBADR;
```

Comment:
why not returning `errno` if `origin` is not "linux"? this will be backward compatible.

Reply:
because other OSes will assign `errno` numbers differently. 

The logic I implemented uses the errno number under one of two conditions:

1. if the origin is specified and is "linux"
2. if no origin is specified

it will *not* use it if:

3. if the origin is specified and is not "linux".

Thus if people one day intend to implement this same logic on other OSes it's sufficient to set the origin properly, and we'll not use the numeric number on linux.

and it also retains compat with the status quo ante: if the field is not set we basically assume linux and use the number as we always did.